fish-rust/src/input_common.rs

```diff
@@ -1,21 +1,586 @@
+            return;
+        };
+        let last = queue
+            .range(first..)
+            .position(is_char)
+            .map_or(queue.len(), |x| x + first);
+        // Move the non-char events to the front, retaining their order.
+        let elems: Vec<CharEvent> = queue.drain(first..last).collect();
+        for elem in elems.into_iter().rev() {
+            queue.push_front(elem);
+        }
+    }
+
+    /// Add multiple readline events to the front of the queue of unread characters.
+    /// The order of the provided events is not changed, i.e. they are not inserted in reverse
+    /// order. That is, the first element in evts will be the first element returned.
+    fn insert_front<I>(&mut self, evts: I)
+    where
+        I: IntoIterator<Item = CharEvent>,
+        I::IntoIter: DoubleEndedIterator,
+    {
+        let queue = self.get_queue_mut();
+        let iter = evts.into_iter().rev();
+        queue.reserve(iter.size_hint().0);
+        for evt in iter {
+            queue.push_front(evt);
+        }
+    }
+
+    /// Forget all enqueued readline events in the front of the queue.
+    fn drop_leading_readline_events(&mut self) {
+        let queue = self.get_queue_mut();
+        while let Some(evt) = queue.front() {
+            if evt.is_readline() {
+                queue.pop_front();
+            } else {
+                break;
+            }
+        }
+    }
+
+    /// Override point for when we are about to (potentially) block in select(). The default does
+    /// nothing.
+    fn prepare_to_select(&mut self) {}
+
+    /// Override point for when when select() is interrupted by a signal. The default does nothing.
+    fn select_interrupted(&mut self) {}
+
+    /// Override point for when when select() is interrupted by the universal variable notifier.
+    /// The default does nothing.
+    fn uvar_change_notified(&mut self) {}
+
+    /// \return if we have any lookahead.
+    fn has_lookahead(&self) -> bool {
+        !self.get_queue().is_empty()
+    }
+}
+
+/// A simple, concrete implementation of InputEventQueuer.
+pub struct InputEventQueue {
+    queue: VecDeque<CharEvent>,
+    in_fd: RawFd,
+}
+
+impl InputEventQueue {
+    pub fn new(in_fd: RawFd) -> InputEventQueue {
+        InputEventQueue {
+            queue: VecDeque::new(),
+            in_fd,
+        }
+    }
+}
+
+impl InputEventQueuer for InputEventQueue {
+    fn get_queue(&self) -> &VecDeque<CharEvent> {
+        &self.queue
+    }
+
+    fn get_queue_mut(&mut self) -> &mut VecDeque<CharEvent> {
+        &mut self.queue
+    }
+
+    fn get_in_fd(&self) -> RawFd {
+        self.in_fd
+    }
+}
+
+#[test]
+fn test_push_front_back() {
+    let mut queue = InputEventQueue::new(0);
+    queue.push_front(CharEvent::from_char('a'));
+    queue.push_front(CharEvent::from_char('b'));
+    queue.push_back(CharEvent::from_char('c'));
+    queue.push_back(CharEvent::from_char('d'));
```

Comment:
not so important here but when I'm navigating the call graph, I almost always want
to ignore test code. Separate files makes that easy